### PR TITLE
Extended bool filter to support python truthiness (#21914).

### PR DESCRIPTION
**Pull request by anryko**
##### ISSUE TYPE

 - Feature Pull Request

##### ANSIBLE VERSION

```
ansible 2.3.0 (filter_bool_python 5904636d35) last updated 2017/02/27 23:23:58 (GMT +200)
```

##### SUMMARY

Discussed in #21914
Adds functionality to utilize python truthiness while using ```|bool``` filter by providing filter style choice in a similar way as in ```|comment``` filter.

Playbook example:
```yaml
- hosts: localhost
  tasks:
    - set_fact:
        yaml_dict_empty: "{{ {}|bool }}"
        yaml_dict_not_empty: "{{ {'key': 1}|bool }}"
        yaml_str_empty: "{{ ''|bool }}"
        yaml_str_not_empty: "{{ 'aaa'|bool }}"
        yaml_str_eq_yes: "{{ 'yes'|bool }}"
        yaml_int_0: "{{ 0|bool }}"
        yaml_int_1: "{{ 1|bool }}"
        yaml_int_negative: "{{ -7|bool }}"
        python_dict_empty: "{{ {}|bool('python') }}"
        python_dict_not_empty: "{{ {'key': 1}|bool('python') }}"
        python_str_empty: "{{ ''|bool('python') }}"
        python_str_not_empty: "{{ 'aaa'|bool('python') }}"
        python_str_eq_yes: "{{ 'yes'|bool('python') }}"
        python_int_0: "{{ 0|bool('python') }}"
        python_int_1: "{{ 1|bool('python') }}"
        python_int_negative: "{{ -7|bool('python') }}"
    - debug:
        var: hostvars[inventory_hostname]
```
Result:
```bash
        "python_dict_empty": false,
        "python_dict_not_empty": true,
        "python_int_0": false,
        "python_int_1": true,
        "python_int_negative": true,
        "python_str_empty": false,
        "python_str_eq_yes": true,
        "python_str_not_empty": true,
        "yaml_dict_empty": false,
        "yaml_dict_not_empty": false,
        "yaml_int_0": false,
        "yaml_int_1": true,
        "yaml_int_negative": false,
        "yaml_str_empty": false,
        "yaml_str_eq_yes": true,
        "yaml_str_not_empty": false
```

### CODE REVIEW

**Comment by bcoca** [comment]
lgtm

**Comment by joejulian** [comment]
:+1:

File: lib/ansible/plugins/filter/core.py [hunk 1]
```diff
@@ -114,15 +114,20 @@ def to_nice_json(a, indent=4, *args, **kw):
...
-        return True
-    return False
+    if style == 'python':
+        return bool(a)
+    elif style == 'yaml':
+        if a is None or isinstance(a, bool):
+            return a
+        if isinstance(a, string_types):
+            a = a.lower()
+        if a in ('yes', 'on', '1', 'true', 1):
```
Comment (bcoca) [inline lib/ansible/plugins/filter/core.py:126]: this should probably use the BOOLEAN tests from module_utils

**Comment by prutseltje** [comment]
Are there any plans to merge this? It would be a nice feature to have!

**Comment by guoqiao** [comment]
One more thing,  in the original code, if a is None, it still return None, other than False.
Should we fix that as well? This filter is called `bool`,  it should always return boolean value.

**Comment by guoqiao** [comment]
This PR is good,  and I notice ansible is using this approach for many cases. It just worries me that the interface of ansible is becoming more complicated, and ugly.
Also, this bool case is different from other ones: the original code is actually wrong, instead of keep it and extend it, I would prefer to fix it.

**Comment by corbinstuard** [comment]
Is there anything holding up this PR? I would also greatly appreciate it.

**Comment by Akasurde** [comment]
@anryko Are you still working on this ? Let me know, thanks.

**Comment by anryko** [comment]
@Akasurde if there is a formed opinion about backward compatible and non breaking improvements I could introduce to this PR, I would be happy to contribute. I'm not working on it for a very long time now :).

**Comment by Akasurde** [comment]
Closing this since #62602 is merged. Thanks for the contribution.

**Pull request closed without merging**